### PR TITLE
SCons: Drop support for Python 2

### PR DESCRIPTION
**Pull request by akien-mga**
We now require SCons 3.0+ (first version with Python 3 support),
and we set min required Python 3 version to 3.5 (3.4 and earlier are
EOL).

--

https://docs.godotengine.org/en/latest/development/compiling/ docs should be updated accordingly to remove references to Python 2.

If we decide to use f-strings to simplify our code builders (might be a good idea), we should bump the min Python 3 version to 3.6.

### CODE REVIEW

**Comment by reduz** [comment]
That will make Godot not as easy to use or build from source in a lot of distros, specially distros like Ubuntu which are meant to be user friendly and where a lot of users may not want to install scons manually (or are just not used to install software manually).

**Comment by Calinou** [comment]
@reduz As far as I know, you can still use the system-provided SCons on Ubuntu 18.04. This isn't the case on Ubuntu 16.04 though, but it'll be EOL soon enough.

**Comment by akien-mga** [comment]
> As far as I know, you can still use the system-provided SCons on Ubuntu 18.04.

No, Ubuntu 18.04's `scons` package is the Python 2 version.

Users with an outdated distro package with no Python 3 module will have to use pip:
```
pip3 install --user scons
```
or
```
sudo pip3 install scons
```

**Comment by akien-mga** [comment]
I added instructions for pip usage in godotengine/godot-docs#3309.

If users of other platforms struggle with it, we might need to add more details instructions on how to install a valid SCons version in "Introduction to the buildsystem", and reference it on all pages in lieu of the link to the SCons homepage.

**Comment by akien-mga** [comment]
For the reference, I checked and Ubuntu up to 19.10 still provided the Python 2 flavor in their `scons` package (-_-)... But 20.04 should come with the Python 3 one: https://packages.ubuntu.com/focal/scons

File: SConstruct [hunk 1]
```diff
@@ -13,7 +13,12 @@ import methods
 import gles_builders
 from platform_methods import run_in_subprocess
 
-# scan possible build platforms
+# We now require Python 3.5+
+if not (sys.version_info.major == 3 and sys.version_info.minor >= 5):
+    print("Godot's buildsystem requires Python 3.5 or later. Be sure to also install SCons for Python 3.")
+    Exit(1)
+
```
Comment (Xrayez) [inline SConstruct]: I think [`EnsurePythonVersion(3, 5)`](https://scons.org/doc/production/HTML/scons-man.html#f-EnsurePythonVersion) can be used instead provided by SCons next to `EnsureSConsVersion`.

Comment (akien-mga) [reply SConstruct]: Thanks, done.

File: SConstruct [hunk 2]
```diff
@@ -13,7 +14,7 @@ import methods
 import gles_builders
 from platform_methods import run_in_subprocess
 
-# scan possible build platforms
+# Scan possible build platforms
```
Comment (aaronfranke) [inline SConstruct]: Could also take the opportunity to capitalize other comments, and add periods to the comments.

**Comment by aaronfranke** [comment]
Perhaps it would be best if this PR waits until after Ubuntu 20.04 becomes commonplace? A good indicator would be whenever Travis CI gets support for Ubuntu 20.04, and we can set Travis to that.

File: SConstruct (hunk 2 above)
Comment (akien-mga) [reply SConstruct]: No, cosmetic changes in unrelated parts of the file are bad practice.

**Comment by akien-mga** [comment]
> Perhaps it would be best if this PR waits until after Ubuntu 20.04 becomes commonplace? 

I don't think we should limit ourselves for Ubuntu specifically. It's trivial to install SCons with Python 3 support using pip, which is available on all supported Ubuntu versions.

> A good indicator would be whenever Travis CI gets support for Ubuntu 20.04, and we can set Travis to that.

That would be a pretty bad indicator actually, as Travis is not known for being quick to provide new images. Ubuntu 16.04 became their default platform [in April 2019](https://blog.travis-ci.com/2019-04-15-xenial-default-build-environment), and it had only started being available in beta in [November 2018](https://blog.travis-ci.com/2018-11-08-xenial-release).

But again, this is a change in the `master` branch, which is currently unstable and only meant to be used by contributors, who can manage to install the appropriate SCons version on any system.

And it's time Python 2 to seriously go away, so as downstream users we also need to push in that direction, even if it means that some users will have to use pip instead of apt...

File: SConstruct (hunk 2 above)
Comment (akien-mga) [reply SConstruct]: Well it *used* to be in the same code chunk as I had added code right below it, but that's no longer the case now.

**Comment by reduz** [comment]
So far it's not trivial to run scons with python3 in Ubuntu:
* pip3 needs to be installed manually
* My system has been upgraded a few times, and due to old .bashrc ~/.local/bin is not in PATH, so only installing as root works
* but it does not really work, because the "python" executable is "python2", so scons will fail to run. Updating this alternative in older Ubuntu causes "apt" to break. I had to harcode the scons binary to use python3.

If this can't be solved without hacks, I think we will need to revert this PR.

**Comment by akien-mga** [comment]
I'm pretty sure the second binary installed from pip should use the same python version as pip, not rely on what "python" is in shell.

Are you sure it's not picking up /usr/bin/scons from the Ubuntu scons package? (Which I would advise to uninstall.)

**Comment by akien-mga** [comment]
You can also file a Ubuntu bug report to ask them to provide a `python3-scons` package like the RPM world has done for 2 years now before defaulting their `scond` to py3.

**Comment by akien-mga** [comment]
> but it does not really work, because the "python" executable is "python2", so scons will fail to run. Updating this alternative in older Ubuntu causes "apt" to break. I had to harcode the scons binary to use python3.

I tested in an Ubuntu 14.04 container, and indeed, the `/usr/local/bin/scons` script uses `/usr/bin/env python` so it doesn't work. I'll see with upstream SCons if they could fix that in their pip distribution.

File: SConstruct [hunk 3]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python
```
Comment (madmiraal) [inline SConstruct]: You could try changing the shebang to `#!/usr/bin/env python3`.

Comment (akien-mga) [reply SConstruct]: This shebang is mostly to help IDEs with auto-completion, we don't actually run SConstruct via Python, it's parsed by SCons itself.

Comment (madmiraal) [reply SConstruct]: True. It's the other way around. Changing the shebang at the beginning of the scons script itself to `#! /usr/bin/python3` makes it work (and there is no need to install the latest scons using pip3!)

Comment (Calinou) [reply SConstruct]: If you use `/usr/bin/python3`, this means the user-chosen Python version (e.g. with [pyenv](https://github.com/pyenv/pyenv)) will be ignored :slightly_frowning_face:

Comment (madmiraal) [reply SConstruct]: @Calinou This change is only needed for distributions that still have Python v2 as the default i.e. python points to python2 and for versions of scons that are have the shebang pointing to /usr/bin/python2 as for example my scons version 3.1.1 on Ubuntu was doing.

I'll submit an update to the documentation to reflect this.

**Comment by aaronfranke** [comment]
How are you supposed to do this on Ubuntu? I installed `scons` via `pip3` and I'm having this issue:

```
$ scons
SCons import failed. Unable to find engine files in:
  /home/aaronfranke/.local/lib/scons-3.1.2
  /usr/lib/scons-3.1.2
  /usr/local/lib/scons-3.1.2
  /home/aaronfranke/.local/lib/python2.7/site-packages/scons-3.1.2
  /usr/local/lib/python2.7/site-packages/scons-3.1.2
  /usr/lib/scons-3.1.2
  /home/aaronfranke/.local/lib/scons
  /usr/lib/scons
  /usr/local/lib/scons
  /home/aaronfranke/.local/lib/python2.7/site-packages/scons
  /usr/local/lib/python2.7/site-packages/scons
  /usr/lib/scons
Traceback (most recent call last):
  File "/home/aaronfranke/.local/bin/scons", line 195, in <module>
    import SCons.Script
ImportError: No module named SCons.Script
```

This happens even outside of Godot's folder, I can't run `scons --version`. I've tried both `pip3 install scons` and `sudo pip3 install scons` (after `sudo apt remove scons` obviously), and pip3 does say that I've installed `scons-3.1.2`.

**Comment by akien-mga** [comment]
SCons' pip packaging seems to be a bit unconventional, and they don't use the classical setuptools that would generate a proper shebang. You'll have to edit `/usr/local/bin/scons` or `~/.local/bin/scons` to ensure that it points to `/usr/bin/env python3` and not `python`.

Alternatively, SCons devs recommend running it in a virtual env to alleviate that, but I'm not familiar with it. @Faless is doing that so we'll document it.

**Comment by madmiraal** [comment]
There is no need to install the latest version of SCons via pip3 as long as you're running SCons v3.0+. As per the updated [instructions](https://docs.godotengine.org/en/latest/development/compiling/compiling_for_x11.html), you just need to change the version of Python that SCons uses by changing the shebang (the first line) of the SCons script file to `#! /usr/bin/python3`. Use the command `which scons` to find the location of the SCons script file.

**Comment by aaronfranke** [comment]
Is there an upstream bug report about this? I would expect a package installed via `pip3` to use Python 3 (since `pip3` has `3` in the name, so it's specific to Python 3).

**Comment by madmiraal** [comment]
There is no bug with SCons. SCons v3.0+ supports both Python 2 and Python 3 and SCons v3.1.2 uses your systems default version of Python.

On Ubuntu the default version of Python is v2. In addition the shebang on the version of SCons installed via the `apt-get` has the shebang `#! /usr/bin/python2` which forces SCons to use Python v2. To get around both of these issues you just need to update the shebang on the SCons script you're using to `#! /usr/bin/python3`.

**Comment by aaronfranke** [comment]
@madmiraal If SCons v3.1.2 supports Python 2, why does it fail with `#! /usr/bin/env python` as I've observed above?

This isn't even related to Godot, this happens with `scons` called from my home directory. If running `pip3 install scons`, and then `scons`, doesn't work... how is that not a bug with SCons?

**Comment by Calinou** [comment]
I had to replace the shebang after installing SCons via pip on Ubuntu 16.04 as demonstrated here: https://github.com/Calinou/godot-builds-ci/blob/ff49f8bd67123c70bf14ff5dca913d867665ff9e/scripts/gitlab-ci/before_script.sh#L37-L39

**Comment by madmiraal** [comment]
@aaronfranke When you use pip3, it installs the Python v3 versions of the files it needs in the Python v3 directories. When you run SCons it runs /usr/bin/python. In Ubuntu /usr/bin/python points to /usr/bin/python2. Your system then searches the Python v2 directories for the files and it can't find them.

**Comment by Faless** [comment]
So, the easiest way I found to get SCons with python3 on old distros is using virtualenv.
Assuming you have python3 installed, and the virtualenv module (on ubuntu `apt install python3 python3-venv`).

Create the virtualenv:
`python3 -m venv .venv3`
Will create a virtual env in the `.venv3` folder in the current directory. You only need to run this once

To activate the venv:
`. .venv3/bin/activate` (**note the leading dot**)
You should see `(.venv3) ` leading your terminal line after this command (e.g. `(.venv3) fales@fales-desktop:~$`)

Install scons in the virtual env (after activating it):
`pip install scons` (will automatically use the correct pip version when the environment is active)
Of course, you only need to run this once in the virtualenv.

When the venv is activate (`. .venv3/bin/activate`) the `scons` command will use python3.

To deactivate the virtualenv after activating it for a terminal simply type: `deactivate`.

So, to recap:
```bash
apt install python3 python3-venv # Once, for ubuntu, check your distro package system for virtualenv
python3 -m venv .venv3 # Once
. .venv3/bin/activate # Always
pip install scons # Once
scons --version # Whatever scons command you need to run now!
```

**Comment by akien-mga** [comment]
I guess once the virtual env is set up as described by Fabio, it might make sense to write an alias in your `~/.bashrc`, like:
```
alias scons-3=". .venv3/bin/activate && scons"
```

**Comment by aaronfranke** [comment]
It would make sense for this to be added to the documentation :)

I spoke with a member of the SCons team, this is a known issue and they plan to fix it when they refactor their packaging system for SCons 4.0. He referenced https://github.com/SCons/scons/issues/3451

**Comment by madmiraal** [comment]
I suspect the reason it works is that they've changed the shebang to `#!/usr/bin/python3`.

**Comment by akien-mga** [comment]
BTW, since the same `scons` script seems to work with both `python2` and `python3` and does not rely on any version-specific modules in `site-packages`, another workaround is simply:
```
alias scons="python3 `which scons`"
```

You can put this in your `~/.bashrc` so that `scons` will always be aliases to `python3 /usr/bin/scons` or wherever your scons script is located. I don't think it should trigger a recursive issue, but if you fear that, you can also do:
```
alias scons="python3 /path/to/scons"
```
or use a unique name for the alias.
